<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Customising and running nf-core pipelines – Peter Mac Nextflow Workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-fe31d7bd2deb2a321418d01dba375a64.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a14e3238c51140e99ccc48519b6ed9ce.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed fullcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Peter Mac Nextflow Workshop</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../workshops/00_setup.html"> 
<span class="menu-text">Setup</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-sessions" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Sessions</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-sessions">    
        <li>
    <a class="dropdown-item" href="../sessions/1_intro_run_nf.html">
 <span class="dropdown-text">Day 1: Introduction to Nextflow and Nf-core</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../sessions/2_nf_dev_intro.html">
 <span class="dropdown-text">Day 2: Developing bioinformatics workflows with Nextflow</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-workshop" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Workshop</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-workshop">    
        <li>
    <a class="dropdown-item" href="../workshops/1.1_intro_nextflow.html">
 <span class="dropdown-text">1.1 Introduction to Nextflow</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/1.2_intro_nf_core.html">
 <span class="dropdown-text">1.2 Introduction to nf-core</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/2.1_customise_and_run.html">
 <span class="dropdown-text">2.1 Customising &amp; running nf-core pipelines</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/2.2_troubleshooting.html">
 <span class="dropdown-text">2.2 Troubleshooting Nextflow run</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/1.3_processes_and_channels.html">
 <span class="dropdown-text">3.1 Introduction to processes and channels</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/3.1_creating_a_workflow.html">
 <span class="dropdown-text">3.2 Creating a workflow</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/2.3_tips_and_tricks.html">
 <span class="dropdown-text">3.3 Best practise, tips and tricks</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/5.1_nf_core_template.html">
 <span class="dropdown-text">4.1 Creating workflows using nf-core templates</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/5.2_nf_core_modules.html">
 <span class="dropdown-text">5.1 nf-core modules and subworkflows</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/5.3_metadata_propagation.html">
 <span class="dropdown-text">6.1 Metadata Propagation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../workshops/5.4_pipeline_testing.html">
 <span class="dropdown-text">7.1 Pipeline version control and additional resources</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><strong>Customising and running nf-core pipelines</strong></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<hr>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Run an nf-core pipeline</li>
<li>Specify different parameters to customise the running of an nf-core pipeline</li>
<li>Examine output files from an nf-core pipeline</li>
<li>Customise config files for running an nf-core pipeline</li>
</ul>
</div>
</div>
<section id="environment-setup" class="level3">
<h3 class="anchored" data-anchor-id="environment-setup"><strong>2.1.1. Environment setup</strong></h3>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<strong>Institute specific instructions</strong>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Reminder before starting this session, make sure you follow your institute’s HPC rulebook re: how to run a workflow manager.</p>
<p>At Peter Mac HPC, this would mean not running the nextflow pipeline on the login node, load the nextflow &amp; apptainer modules, and set the container cache location (or use the institutional config)</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>srun --pty -p &lt;PARTITION&gt; --mem 8GB --mincpus 2 -t 0-5:00 bash</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>module load nextflow/24.10.5</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>export NXF_APPTAINER_CACHEDIR="/config/binaries/singularity/containers_devel/nextflow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Before we start, please download the test datasets into your home directory</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>git clone --single-branch --branch rnaseq https://github.com/nf-core/test-datasets ~/rnaseq_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>And for the purpose of this workshop, we are going to cache the containers by running a test (with a profile) to save time for our processing later</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>export NXF_APPTAINER_CACHEDIR="/home/${USER}/apptainer_cache"</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>nextflow run nf-core/rnaseq -r 3.14.0 -profile test,apptainer --outdir ~/test_rnaseq --max_memory 7.GB --max_cpus 2 </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Please also create a folder inside our work directory called <code>lesson2.1</code> and move into it:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>mkdir ./lesson2.1 &amp;&amp; cd $_</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="understanding-an-nf-core-pipeline" class="level3">
<h3 class="anchored" data-anchor-id="understanding-an-nf-core-pipeline"><strong>2.1.2. Understanding an nf-core pipeline</strong></h3>
<p>The following sections of the nf-core documentation can be used to understand what a particular pipeline is doing, to inform your choices about aspects of pipeline-specific customisations. For this section, we will investigate <code>nf-core/rnaseq</code> pipeline.</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Docs</th>
<th>Description</th>
<th>Customisation level</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="https://nf-co.re/rnaseq/3.14.0">Introduction</a></td>
<td>Workflow summary</td>
<td><ul>
<li>workflow</li>
<li>process</li>
</ul></td>
</tr>
<tr class="even">
<td><a href="https://nf-co.re/rnaseq/3.14.0/usage">Usage</a></td>
<td>Inputs and options</td>
<td><ul>
<li>workflow</li>
<li>process</li>
</ul></td>
</tr>
<tr class="odd">
<td><a href="https://nf-co.re/rnaseq/3.14.0/parameters">Parameters</a></td>
<td>Available flags</td>
<td><ul>
<li>workflow</li>
<li>process</li>
<li>compute resources</li>
</ul></td>
</tr>
<tr class="even">
<td><a href="https://nf-co.re/rnaseq/3.14.0/output">Output</a></td>
<td>Files from all processes processes</td>
<td><ul>
<li>workflow</li>
<li>process</li>
<li>tool</li>
</ul></td>
</tr>
</tbody>
</table>
<section id="pipeline-structure" class="level4">
<h4 class="anchored" data-anchor-id="pipeline-structure"><strong>Pipeline structure</strong></h4>
<p>Looking at the nf-core/rnaseq pipeline structure provided in the <a href="https://nf-co.re/rnaseq/3.14.0">introduction</a>, we can see that the developers have:</p>
<ol type="1">
<li>Organised the workflow into 5 stages based on the type of work that is being done</li>
<li>Provided a choice of multiple methods and specified defaults</li>
<li>Provided a choice of tool for some steps</li>
</ol>
<p><img src="./media/2.1_pipeline-choice.png" class="img-fluid"></p>
<p><strong>Quiz</strong>: Observing the diagram above, which statement is true regarding the choice of alignment and quantification methods provided by the nf-core/rnaseq pipeline?</p>
<p><strong>A.</strong> The pipeline uses a fixed method for read alignment and quantification.<br>
<strong>B.</strong> Users can choose between several different methods for read alignment and quantification.<br>
<strong>C.</strong> The pipeline always performs read alignment and quantification using STAR or HISAT2.<br>
<strong>D.</strong> The choice of alignment and quantification method is determined automatically based on the input data.</p>
<div class="callout callout-style-simple callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The correct answer is <strong>B</strong>. The <code>nf-core/rnaseq</code> pipeline allows users to choose between pseudo-alignment and quantification, or genome-based read alignment and quantification.</p>
<ul>
<li><strong>A.</strong> is incorrect because the pipeline is not limited to a single method.<br>
</li>
<li><strong>C.</strong> is incorrect because while read alignment and quantification using STAR is the default method, users can also choose the pseudo-alignment.</li>
<li><strong>D.</strong> is also incorrect, as the pipeline only accepts FASTQ files as input, and the choice of alignment and quantification method must be specified by the user.</li>
</ul>
</div>
</div>
</div>
</section>
</section>
<section id="default-pipeline-usage" class="level3">
<h3 class="anchored" data-anchor-id="default-pipeline-usage"><strong>Default pipeline usage</strong></h3>
<p>Typically, nf-core pipelines at a minimum require users to specify a <a href="https://nf-co.re/rnaseq/3.14.0/usage#samplesheet-input">sample sheet</a> (<code>--input</code>) detailing the path to your sample data and any relevant metadata. Additionally, if a reference file version is not provided (using the <code>--genome</code> parameter), a default <a href="https://github.com/nf-core/rnaseq/blob/master/conf/igenomes.config">iGenomes</a> one will be used.</p>
<p>You can see the recommended (typical) run command and all the parameters available for the nf-core/rnaseq pipeline by running:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>nextflow run nf-core/rnaseq -r 3.14.0 --help </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The typical or recommended run command for this <code>rnaseq</code> pipeline is provided at the top of the output:</p>
<p><img src="./media/2.1_default-command.png" class="img-fluid"></p>
<p>It outlines a requirement for:</p>
<ul>
<li><code>--input</code>: An input samplesheet that contains the data to be processed</li>
<li><code>--outdir</code>: A location to store outputs</li>
<li><code>--genome</code>: Relevant reference data</li>
<li><code>-profile</code>: A software management method</li>
</ul>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<strong>Reminder: hyphens matter in Nextflow!</strong>
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Nextflow</strong>-specific parameters use one (<code>-</code>) hyphen, whereas <strong>pipeline</strong>-specific parameters use two (<code>--</code>). In the typical run command above <code>-profile</code> is a <strong>Nextflow</strong> parameter, while <code>--input</code> is a <strong>Nextflow</strong> parameter.</p>
</div>
</div>
</section>
<section id="setting-up-the-run-command" class="level3">
<h3 class="anchored" data-anchor-id="setting-up-the-run-command"><strong>2.1.3. Setting up the run command</strong></h3>
<p>Most of us will need to adjust the default run command for our experiments. Today we’ll be adjusting the typical nf-core/rnaseq run command by:</p>
<ol type="1">
<li>Creating a samplesheet csv, based on the requirements of the pipeline</li>
<li>Providing our own reference files</li>
<li>Using the <code>apptainer</code> software management profile</li>
<li>Customising the execution of some processes</li>
<li>Specifying the computing resource limitations of our session (2 CPUs, 8 GB RAM)</li>
</ol>
<section id="required-parameter---input" class="level4">
<h4 class="anchored" data-anchor-id="required-parameter---input"><strong>Required parameter: <code>--input</code></strong></h4>
<p>We will create a <code>samplesheet.csv</code> based on the documentation provided by <a href="https://nf-co.re/rnaseq/3.14.0/docs/usage/">nf-core/rnaseq</a>.</p>
<p><img src="./media/2.1_rnaseq_samplesheet.png" class="img-fluid"></p>
<p>For <code>rnaseq</code>, a <code>sample</code> value that specifies the sample name, path to FASTQ files, and <code>strandedness</code> is required.</p>
<p>Since we are only testing the pipeline in this session, we only need to work with a couple of samples.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sample,fastq_1,fastq_2,strandedness</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>SRR6357070,../rnaseq_data/testdata/GSE110004/SRR6357070_1.fastq.gz,../rnaseq_data/testdata/GSE110004/SRR6357070_2.fastq.gz,forward</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>SRR6357071,../rnaseq_data/testdata/GSE110004/SRR6357071_1.fastq.gz,../rnaseq_data/testdata/GSE110004/SRR6357071_2.fastq.gz,forward</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="required-parameter---outdir" class="level4">
<h4 class="anchored" data-anchor-id="required-parameter---outdir"><strong>Required parameter: <code>--outdir</code></strong></h4>
<p>Most nf-core pipelines will require user to specified an output directory to dump all the output files to.</p>
<p>For this exercise, we will set our output directory to:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>    --outdir ./rnaseq_small_test_outdir</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="required-input-reference-data" class="level4">
<h4 class="anchored" data-anchor-id="required-input-reference-data"><strong>Required input: reference data</strong></h4>
<p>Many nf-core pipelines have a minimum requirement for reference data inputs. The input reference data requirements for this pipeline are provided in the <a href="https://nf-co.re/rnaseq/3.14.0/docs/usage/#reference-genome-options">usage documentation</a>:</p>
<p><img src="./media/2.1_reference_files.png" class="img-fluid"></p>
<p>In the documentation, we see that the recommended method to provide reference files is to explicitly state them using the <code>--fasta</code> and <code>--gtf</code> parameters. This means we can replace the <code>--genome</code> flag in the typical run command with our own files. To see all available reference file parameters, rerun the pipeline’s help command to view all the available parameters:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>nextflow run nf-core/rnaseq -r 3.14.0 --help</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>From the <strong><code>Reference genome options</code></strong> parameters, we will provide our own files using:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>  --fasta ../rnaseq_data/reference/genome.fasta</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  --gtf ../rnaseq_data/reference/genes.gtf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="optional-parameters" class="level3">
<h3 class="anchored" data-anchor-id="optional-parameters"><strong>Optional parameters</strong></h3>
<p>Now that we have prepared our input and reference data, we will further customise the typical run command by:</p>
<ol type="1">
<li>Using Nextflow’s <code>-profile</code> parameter to specify the <code>apptainer</code> profile</li>
<li>Adding additional process-specific flags to <a href="https://nf-co.re/rnaseq/3.14.0/parameters#skip_alignment">skip alignment</a> and only use <a href="https://nf-co.re/rnaseq/3.14.0/parameters#pseudo_aligner">pseudo-aligner</a></li>
<li>Adding additional max resource flags to specify the <a href="https://nf-co.re/rnaseq/3.14.0/parameters#max_cpus">number of CPUs</a> and <a href="https://nf-co.re/rnaseq/3.14.0/parameters#max_memory">amount of memory</a> available to the pipeline.</li>
</ol>
<p>Using the command line, the following parameters can be set:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>--skip_alignment true</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>--pseudo_aligner salmon</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Inside the <a href="https://github.com/nf-core/rnaseq/blob/3.14.0/nextflow.config"><code>nextflow.config</code></a>, <code>max_memory</code> and <code>max_cpus</code> have been set to the following:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>    max_memory                 = '128.GB'</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    max_cpus                   = 16</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Using the command line, this can be changed with:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>--max_memory '6.GB'</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>--max_cpus 2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="running-the-pipeline" class="level3">
<h3 class="anchored" data-anchor-id="running-the-pipeline"><strong>2.1.4. Running the pipeline</strong></h3>
<p>Putting together all the input &amp; parameters that we specified above, the final command will contain our software profile, input samplesheet, output directory, reference files, custom pipeline steps, and custom resources to use.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>nextflow run nf-core/rnaseq -r 3.14.0 \</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    -profile apptainer \</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    --input samplesheet.csv \</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    --outdir ./rnaseq_small_test_outdir \</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    --fasta ../rnaseq_data/reference/genome.fasta \</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    --gtf ../rnaseq_data/reference/genes.gtf \</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    --skip_alignment true \</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    --pseudo_aligner salmon \</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    --max_memory '6.GB' \</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    --max_cpus 2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can see how far we’ve customised the typical run command from the original example of :</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>nextflow run nf-core/rnaseq --input samplesheet.csv --genome GRCh37 -profile docker</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now that we have prepared our data and chosen which parameters to apply, run the pipeline using the customised command we created above. Take a look at the <code>stdout</code> printed to the screen. Your workflow configuration and parameter customisations are all documented here. You can use this to confirm if your parameters have been correctly passed to the run command:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./media/2.1_nf-core-stdout.png" class="img-fluid figure-img"></p>
<figcaption>Note that this screenshot is only demonstrating the different sections, not necessary the parameters that we have specified above</figcaption>
</figure>
</div>
<p>As the workflow starts, you will also see a number of processes that are created underneath this. Recall that processes are executed <strong>independently</strong> and can run in <strong>parallel</strong>. Nextflow manages the data dependencies between processes, ensuring that each process is executed only when its input data is available, and all of its dependencies have been satisfied.</p>
<p>To understand how this is coordinated, consider the <code>STAR_ALIGN</code> process.</p>
<p><img src="./media/2.1_nf-core_processes.png" class="img-fluid"></p>
<p>Notice a few things:</p>
<ul>
<li>We can see which inputs are being processed by looking at the end of the process name</li>
<li>When a process starts it progressively spawns tasks for all inputs to be processed</li>
<li>Two <code>TRIMGALORE</code> processes are created, one for each sample in our <code>samplesheet.csv</code>. This process has to complete before <code>STAR_ALIGN</code> begins</li>
<li>Once a <code>TRIMGALORE</code> task is completed for a sample, the <code>STAR_ALIGN</code> task for that sample begins</li>
<li>When the <code>STAR_ALIGN</code> process starts, it spawns 2 tasks, one for each sample in our samplesheet</li>
</ul>
<p><strong>Exercise</strong></p>
<p>While we can specify parameters to a pipeline using the command line, this can be messy and result in huge <code>nextflow run</code> commands where the parameters we used is not documented. Recall <a href="../workshops/1.2_intro_nf_core.html#parameter-files">earlier</a> that a <code>-params-file</code> Nextflow parameter can be used to supply parameters to the pipeline.</p>
<p>Your task: Create a a parameter file <code>workshop-params.yaml</code>, that contains our customised <strong>pipeline</strong> parameters. How can this file then be used in the <code>nextflow run</code> command? (<em>Note</em>: <strong>Nextflow</strong> parameters can’t be supplied inside a parameter file)</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The parameter file <code>workshop-params.yaml</code> should contain the following:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>input: "./samplesheet.csv"</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>outdir: "./rnaseq_small_test_outdir"</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>fasta: "../rnaseq_data/reference/genome.fasta"</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>gtf: "../rnaseq_data/reference/genes.gtf"</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>skip_alignment: true</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>pseudo_aligner: "salmon"</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>max_memory: "6.GB"</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>max_cpus: 2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Note that here the <strong>full</strong> path to our reference files is provided. Since <code>-profile</code> is a <strong>Nextflow</strong> parameter and not a <strong>pipeline</strong> parameter, it’s not listed in the parameter file.</p>
<p>To run the pipeline using this parameter file, the following command can be used:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>nextflow run nf-core/rnaseq -r 3.14.0 \</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    -profile apptainer \</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    -params-file ./workshop-params.yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="examine-the-outputs" class="level4">
<h4 class="anchored" data-anchor-id="examine-the-outputs"><strong>Examine the outputs</strong></h4>
<p>Once your pipeline has completed, you should see this message printed to your terminal:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>-[nf-core/rnaseq] Pipeline completed successfully -</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>Completed at: 27-May-2025 15:25:54</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>Duration    : 2m 15s</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>CPU hours   : 0.1</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>Succeeded   : 20</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The pipeline ran successfully.</p>
<p>In the meantime, list (<code>ls -a</code>) the contents of your directory, and you’ll see new directories (and a hidden directories/files) have been created:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>.  ..  .nextflow  .nextflow.log  .nextflow.log.1  rnaseq_small_test_outdir  samplesheet.csv  work  workshop-params.yaml</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Nextflow has created 2 new output directories, <strong>work</strong> and <strong>rnaseq_small_test_outdir</strong> in the current directory.</p>
<ul>
<li><p><strong>The <code>work</code> directory</strong>:</p>
<ul>
<li>As each job is ran, a unique sub-directory is created inside the <code>work</code> directory.</li>
<li>These directories house temporary files and various command logs created by a process. This contains all the information required when troubleshooting a failed process.</li>
</ul>
<p>We will talk in more detail about pipeline troubleshooting later in the next section.</p></li>
<li><p><strong>The <code>rnaseq_small_test_outdir</code> directory</strong></p>
<ul>
<li>All final outputs will be presented in a directory specified by the <code>--outdir</code> parameter.</li>
<li>Inside this directory, you should have the output files grouped into common tools:</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>.  ..  fastqc  multiqc  pipeline_info  salmon  trimgalore</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<strong>Key points</strong>
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>nf-core pipelines contain default settings and required inputs that can be customised.</li>
<li>An nf-core pipeline’s Usage, Output, and Parameters documentation can be used to design a suitable run command.</li>
<li>Parameters can be used to customise the workflow, processes, tools, and compute resources.<br>
</li>
</ul>
</div>
</div>
<p><br></p>
</section>
</section>
<section id="next-chapter-troubleshooting-a-nextflow-pipeline-run" class="level3">
<h3 class="anchored" data-anchor-id="next-chapter-troubleshooting-a-nextflow-pipeline-run"><strong>Next Chapter: <a href="../workshops/2.2_troubleshooting.html">Troubleshooting a nextflow pipeline run</a></strong></h3>
<p><br></p>
<hr>
<p>This workshop is adapted from various nextflow training materials, including:</p>
<ul>
<li><a href="https://training.nextflow.io">Nextflow Training Materials</a></li>
<li><a href="https://sydney-informatics-hub.github.io/customising-nfcore-workshop">Customising Nf-Core Workshop</a></li>
<li><a href="https://sydney-informatics-hub.github.io/hello-nextflow/">Hello Nextflow Workshop</a></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/PMCC-BioinformaticsCore\.github\.io\/nextflow-nf-core-workshop-2025_v2\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>