<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="quarto-1.7.29" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />


<title>Nextflow Development - Samplesheet Parsing and nf-core modules – Peter Mac Nextflow Workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<!-- htmldependencies:E3FAD763 -->
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="/index.html">
    <span class="navbar-title">Peter Mac Nextflow Workshop</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse"
  aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation"
  onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="/index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="/workshops/00_setup.html"> 
<span class="menu-text">Setup</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-sessions" role="link" data-bs-toggle="dropdown" aria-expanded="false" >
 <span class="menu-text">Sessions</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-sessions">    
        <li>
    <a class="dropdown-item" href="/sessions/1_intro_run_nf.html">
 <span class="dropdown-text">Day 1: Introduction to Nextflow and Nf-core</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/sessions/2_nf_dev_intro.html">
 <span class="dropdown-text">Day 2: Developing bioinformatics workflows with Nextflow</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-workshop" role="link" data-bs-toggle="dropdown" aria-expanded="false" >
 <span class="menu-text">Workshop</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-workshop">    
        <li>
    <a class="dropdown-item" href="/workshops/1.1_intro_nextflow.html">
 <span class="dropdown-text">1.1 Introduction to Nextflow</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/1.2_intro_nf_core.html">
 <span class="dropdown-text">1.2 Introduction to nf-core</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/2.1_customise_and_run.html">
 <span class="dropdown-text">2.1 Customising &amp; running nf-core pipelines</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/2.2_troubleshooting.html">
 <span class="dropdown-text">2.2 Troubleshooting Nextflow run</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/1.3_processes_and_channels.html">
 <span class="dropdown-text">3.1 Introduction to processes and channels</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/3.1_creating_a_workflow.html">
 <span class="dropdown-text">3.2 Creating a workflow</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/2.3_tips_and_tricks.html">
 <span class="dropdown-text">3.3 Best practise, tips and tricks</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/5.1_nf_core_template.html">
 <span class="dropdown-text">4.1 Creating workflows using nf-core templates</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/5.2_nf_core_modules.html">
 <span class="dropdown-text">5.1 nf-core modules and subworkflows</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/5.3_samplesheet_parsing.html">
 <span class="dropdown-text">6.1 Samplesheet parsing and nf-core modules</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/5.4_module_args_custom_module.html">
 <span class="dropdown-text">6.3 Module arguments and custom modules</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="/workshops/5.4_pipeline_testing.html">
 <span class="dropdown-text">7.1 Pipeline version control and additional resources</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <div id="quarto-toc-target"></div>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><strong>Nextflow Development - Samplesheet Parsing and nf-core modules</strong></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Gain an understanding of how to use nf-core modules in a workflow script</li>
<li>Manipulate and proprogate sample metadata throughout the workflow</li>
</ul>
</div>
</div>
<section id="samplesheet-parsing" class="level2">
<h2><strong>6.1 Samplesheet parsing</strong></h2>
<p>In the <code>./nf-core-customrnaseq/main.nf</code> script, the <code>PIPELINE_INITIALISATION</code> subworkflow created by default from the nf-core template will output a channel that contains the parsed <code>--input</code> samplesheet. This channel is then input into <code>NFCORE_CUSTOMRNASEQ</code>, which launches our analysis workflow containing the newly included modules <code>salmon/index</code> and <code>salmon/quant</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>...</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>include { CUSTOMRNASEQ  } from &#39;./workflows/customrnaseq&#39;</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>include { PIPELINE_INITIALISATION } from &#39;./subworkflows/local/utils_nfcore_customrnaseq_pipeline&#39;</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>...</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>workflow NFCORE_CUSTOMRNASEQ {</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    take:</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    samplesheet // channel: samplesheet read in from --input</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    // WORKFLOW: Run pipeline</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>    CUSTOMRNASEQ (</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>        samplesheet</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    emit:</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    multiqc_report = CUSTOMRNASEQ.out.multiqc_report // channel: /path/to/multiqc_report.html</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>...</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>workflow {</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>    // WORKFLOW: Run main workflow</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>    NFCORE_CUSTOMRNASEQ (</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>        PIPELINE_INITIALISATION.out.samplesheet</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>How does the <code>PIPELINE_INITIALISATION</code> parse the samplesheet?</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
<code>./nf-core-customrnaseq/subworkflows/local/utils_nfcore_customrnaseq_pipeline/main.nf</code>
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb2"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>//</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>// Create channel from input file provided through params.input</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>//</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>workflow PIPELINE_INITIALISATION {</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    take:</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    version           // boolean: Display version and exit</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    validate_params   // boolean: Boolean whether to validate parameters against the schema at runtime</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    monochrome_logs   // boolean: Do not use coloured log outputs</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    nextflow_cli_args //   array: List of positional nextflow CLI args</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    outdir            //  string: The output directory where the results will be saved</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    input             //  string: Path to input samplesheet</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    Channel</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>        .fromList(samplesheetToList(params.input, &quot;${projectDir}/assets/schema_input.json&quot;))</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>        .map {</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>            meta, fastq_1, fastq_2 -&gt;</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                if (!fastq_2) {</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                    return [ meta.id, meta + [ single_end:true ], [ fastq_1 ] ]</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>                } else {</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>                    return [ meta.id, meta + [ single_end:false ], [ fastq_1, fastq_2 ] ]</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>                }</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>        .groupTuple()</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>        .map { samplesheet -&gt;</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>            validateInputSamplesheet(samplesheet)</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>        .map {</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>            meta, fastqs -&gt;</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>                return [ meta, fastqs.flatten() ]</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>        .set { ch_samplesheet }</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>    emit:</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>    samplesheet = ch_samplesheet</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>    versions    = ch_versions</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
</div>
</div>
<p>The important pieces of information that we will explore further in this section are:</p>
<ol type="1">
<li>The use of a <code>schema_input.json</code> to validate the samplesheet metadata</li>
<li>The use of <code>.map { }</code> and <code>.groupTuple()</code> functions to manipulate sample metadata</li>
</ol>
<section id="default-samplesheet-channel" class="level3">
<h3><strong>6.1.1 Default samplesheet channel</strong></h3>
<p>The samplesheet is automatically parsed, resulting in a channel that contains all relevant data specified in the <code>--input</code> parmeter. What does this channel contain?</p>
<p>Open the analysis workflow file <code>workflows/customrnaseq.nf</code>. Use the <code>.view()</code> function inside the <code>workflow</code> scope to view the <code>ch_samplesheet</code> that has been input to the pipeline:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>workflow CUSTOMRNASEQ {</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    take:</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    ch_samplesheet // channel: samplesheet read in from --input</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    ch_samplesheet.view()</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    ch_versions = Channel.empty()</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_files = Channel.empty()</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Now, rerun the pipeline, ensuring <code>-resume</code> is specified in the <code>nextflow run</code> command. <strong>Note</strong> ebsure you are no longer inside your pipeline folder.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>nextflow run ./nf-core-customrnaseq/main.nf -resume -profile apptainer --input ./samplesheet.csv --outdir output</span></code></pre></div>
<p>The channel should have the following structure:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>[[id:SRR6357070, single_end:false], [/.../rnaseq_data/testdata/GSE110004/SRR6357070_1.fastq.gz, /.../rnaseq_data/testdata/GSE110004/SRR6357070_2.fastq.gz]]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>[[id:SRR6357071, single_end:false], [/.../rnaseq_data/testdata/GSE110004/SRR6357071_1.fastq.gz, /.../rnaseq_data/testdata/GSE110004/SRR6357071_2.fastq.gz]]</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>[[id:SRR6357072, single_end:false], [/.../rnaseq_data/testdata/GSE110004/SRR6357072_1.fastq.gz, /.../rnaseq_data/testdata/GSE110004/SRR6357072_2.fastq.gz]]</span></code></pre></div>
<p>This channel contains three elements, one for each sample type. The first element is a tuple, where the first element is a list that represents the sample metadata. This metadata contains the sample name, stored as <code>id</code>, and if the sample is single-ded, stored as <code>single_end</code>. The second element in this tuple contain the paths to the input FASTQ files.</p>
<p>Let’s see how this relates to our samplesheet:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sample,fastq_1,fastq_2</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>SRR6357070,/.../rnaseq_data/testdata/GSE110004/SRR6357070_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357070_2.fastq.gz</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>SRR6357071,/.../rnaseq_data/testdata/GSE110004/SRR6357071_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357071_2.fastq.gz</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>SRR6357072,/.../rnaseq_data/testdata/GSE110004/SRR6357072_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357072_2.fastq.gz</span></code></pre></div>
<p>Notice that the value under the <code>sample</code> column has been assigned as <code>id</code> in the channel metadata. File paths in the <code>fastq_1</code> and <code>fastq_2</code> have been added as the second element in the tuple, which represents the read paths.</p>
<p>This is defined inside the <code>assets/schema_input.json</code> file.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
assets/schema_input.json
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb7"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    &quot;$schema&quot;: &quot;https://json-schema.org/draft/2020-12/schema&quot;,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    &quot;$id&quot;: &quot;https://raw.githubusercontent.com/nf-core/customrnaseq/main/assets/schema_input.json&quot;,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    &quot;title&quot;: &quot;nf-core/customrnaseq pipeline - params.input schema&quot;,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    &quot;description&quot;: &quot;Schema for the file provided with params.input&quot;,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    &quot;type&quot;: &quot;array&quot;,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    &quot;items&quot;: {</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>        &quot;type&quot;: &quot;object&quot;,</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>        &quot;properties&quot;: {</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>            &quot;sample&quot;: {</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>                &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>                &quot;pattern&quot;: &quot;^\\S+$&quot;,</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>                &quot;errorMessage&quot;: &quot;Sample name must be provided and cannot contain spaces&quot;,</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>                &quot;meta&quot;: [&quot;id&quot;]</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>            &quot;fastq_1&quot;: {</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>                &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>                &quot;format&quot;: &quot;file-path&quot;,</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>                &quot;exists&quot;: true,</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>                &quot;pattern&quot;: &quot;^\\S+\\.f(ast)?q\\.gz$&quot;,</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>                &quot;errorMessage&quot;: &quot;FastQ file for reads 1 must be provided, cannot contain spaces and must have extension &#39;.fq.gz&#39; or &#39;.fastq.gz&#39;&quot;</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>            &quot;fastq_2&quot;: {</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>                &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>                &quot;format&quot;: &quot;file-path&quot;,</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>                &quot;exists&quot;: true,</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>                &quot;pattern&quot;: &quot;^\\S+\\.f(ast)?q\\.gz$&quot;,</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>                &quot;errorMessage&quot;: &quot;FastQ file for reads 2 cannot contain spaces and must have extension &#39;.fq.gz&#39; or &#39;.fastq.gz&#39;&quot;</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>        &quot;required&quot;: [&quot;sample&quot;, &quot;fastq_1&quot;]</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
</div>
</div>
<p>In this file, each “property” represents a column that can be present inside the <code>--input</code> samplesheet. If a column name was used inside the samplesheet but didn’t specify it in the <code>assets/schema_input.json</code> file, we will get a warning:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>sample,fastq_1,fastq_2,lane</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>SRR6357070,/.../rnaseq_data/testdata/GSE110004/SRR6357070_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357070_2.fastq.gz,1</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>SRR6357071,/.../rnaseq_data/testdata/GSE110004/SRR6357071_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357071_2.fastq.gz,1</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>SRR6357072,/.../rnaseq_data/testdata/GSE110004/SRR6357072_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357072_2.fastq.gz,1</span></code></pre></div>
<p>Rerunning using the new samplesheet:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>nextflow run ./nf-core-customrnaseq/main.nf -resume -profile apptainer --input ./samplesheet.csv --outdir output</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>WARN: Found the following unidentified headers in /.../lesson4.4/samplesheet.csv:</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>        - lane</span></code></pre></div>
<p>Therefore, if you wish to specify an additional column in the sampleshet (ie. adding sample metadata), the <code>schema_input.json</code> should also be changed to allow for this.</p>
<p>Any required columns are also specified, as the <code>"required"</code> item. Any required items not specified in the samplesheet will cause the pipeline to exit.</p>
<p>For example, removing the <code>SRR6357070_1.fastq.gz</code> file from the samplesheet:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sample,fastq_1,fastq_2</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>SRR6357070,,/.../rnaseq_data/testdata/GSE110004/SRR6357070_2.fastq.gz</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>SRR6357071,/.../rnaseq_data/testdata/GSE110004/SRR6357071_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357071_2.fastq.gz</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>SRR6357072,/.../rnaseq_data/testdata/GSE110004/SRR6357072_1.fastq.gz,/.../rnaseq_data/testdata/GSE110004/SRR6357072_2.fastq.gz</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>!! Only displaying parameters that differ from the pipeline defaults !!</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>* The nf-core framework</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    https://doi.org/10.1038/s41587-020-0439-x</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>* Software dependencies</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    https://github.com/nf-core/customrnaseq/blob/master/CITATIONS.md</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>ERROR ~ Validation of pipeline parameters failed!</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a> -- Check &#39;.nextflow.log&#39; file for details</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>The following invalid input values have been detected:</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>* --input (./samplesheet.csv): Validation of file failed:</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>        -&gt; Entry 1: Missing required field(s): fastq_1</span></code></pre></div>
<p>Inside the <code>"sample"</code> property, the <code>"meta"</code> has been set to <code>["id"]</code>. This is the value in the channel metadata that the sample name will be assigned to. We will investigate this further later in the session.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Many existing nf-core nodules rely on the input metadata having <strong>at least</strong> the <code>id</code> value – it is not recommended to change this name from the default.</p>
</div>
</div>
</section>
<section id="input-channels-to-an-nf-core-module" class="level3">
<h3><strong>6.1.2 Input channels to an nf-core module</strong></h3>
<p>Now that we know the contents of our parsed samplesheet channel, let’s check what inputs are required to our two processes.</p>
<section id="process-salmonindex" class="level4">
<h4>Process <code>salmon/index</code></h4>
<p>From the <code>salmon/index</code> <a href="https://github.com/nf-core/rnaseq/blob/b96a75361a4f1d49aa969a2b1c68e3e607de06e8/modules/nf-core/salmon/index/main.nf">module GitHub page</a>, we see that the process requires two inputs: a <code>genome_fasta</code> file, and a <code>transcript_fasta</code> file.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>process SALMON_INDEX {</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    tag &quot;$transcript_fasta&quot;</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    label &quot;process_medium&quot;</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    conda &quot;${moduleDir}/environment.yml&quot;</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    container &quot;${ workflow.containerEngine == &#39;singularity&#39; &amp;&amp; !task.ext.singularity_pull_docker_container ?</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>        &#39;https://depot.galaxyproject.org/singularity/salmon:1.10.3--h6dccd9a_2&#39; :</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>        &#39;biocontainers/salmon:1.10.3--h6dccd9a_2&#39; }&quot;</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    input:</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    path genome_fasta</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>    path transcript_fasta</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Let’s take a closer look at the main Nextflow script that we use to launch the pipeline. Near the top of the script, it provides an example of how to set parameters. These parameters can be specified to the <code>nextflow run</code> command using a parameter <code>.yaml</code> file, specified with <code>-params-file</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    GENOME PARAMETER VALUES</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>// TODO nf-core: Remove this line if you don&#39;t need a FASTA file</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>//   This is an example of how to use getGenomeAttribute() to fetch parameters</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>//   from igenomes.config using `--genome`</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>params.fasta = getGenomeAttribute(&#39;fasta&#39;)</span></code></pre></div>
<p>In the template, it provides an example of how to set the <code>fasta</code> parameter that can be passed to the workflow. We will need to add an additional <code>transcript_fasta</code> parameter.</p>
<p>Edit that code block to the following:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    GENOME PARAMETER VALUES</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>// TODO nf-core: Remove this line if you don&#39;t need a FASTA file</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>//   This is an example of how to use getGenomeAttribute() to fetch parameters</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>//   from igenomes.config using `--genome`</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>params.fasta = getGenomeAttribute(&#39;fasta&#39;)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>params.transcript_fasta = getGenomeAttribute(&#39;transcript_fasta&#39;)</span></code></pre></div>
<p>The <code>getGenomeAttribute</code> will access the <code>fasta</code> or <code>transcript_fasta</code> that’s specified to it via the command line or a parameter file, and re-assign it as a Nextflow parameter that can be accessed in workflows.</p>
<p>Now, we can create a parameter file <code>params.yaml</code>, and specify the paths to <code>fasta</code> and <code>transcript_fasta</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>fasta: &quot;/.../rnaseq_data/reference/genome.fasta&quot;</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>transcript_fasta: /.../rnaseq_data/reference/transcriptome.fasta&quot;</span></code></pre></div>
<p>Since these parameters are strings, we can convert them into channels using the <code>Channel.fromPath</code> channel factory. Open your analysis workflow script <code>nf-core-customrnaseq/workflows/customrnaseq.nf</code> and add the following:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>workflow CUSTOMRNASEQ {</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    take:</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    ch_samplesheet // channel: samplesheet read in from --input</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    // Gather reference files</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    ch_genome_fasta = Channel.fromPath(params.fasta)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    ch_transcript_fasta = Channel.fromPath(params.transcript_fasta)</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    ch_versions = Channel.empty()</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_files = Channel.empty()</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Nextflow parameters are global variables that can be accessed by any script within the pipeline. Therefore, it doesn’t need to be passed from one file/workflow/process to another.</p>
</div>
</div>
<p>The newly created channels <code>ch_genome_fasta</code> and <code>ch_transcript_fasta</code> match the inputs defined in the <code>SALMON_INDEX</code> process. Suppy these channels as inputs to the <code>SALMON_INDEX</code>, inside <code>workflow CUSTOMRNASEQ { ... }</code></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>    SALMON_INDEX ( </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>        ch_genome_fasta,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>        ch_transcript_fasta</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
<p>Rerun the pipeline, specifying the <code>params.yaml</code> file in the <code>nextflow run</code> command</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>nextflow run ./nf-core-customrnaseq/main.nf -resume -profile apptainer --input ./samplesheet.csv --outdir output -params-file ./params.yaml </span></code></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
stdout
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb20"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a> N E X T F L O W   ~  version 24.10.5</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>Launching `./nf-core-customrnaseq/main.nf` [goofy_cori] DSL2 - revision: 5492b74b7a</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>                                        ,--./,-.</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>        ___     __   __   __   ___     /,-._.--~&#39;</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  |\ | |__  __ /  ` /  \ |__) |__         }  {</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  | \| |       \__, \__/ |  \ |___     \`-._,-`-,</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>                                        `._,._,&#39;</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  nf-core/customrnaseq 1.0.0dev</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>Input/output options</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  input              : ./samplesheet.csv</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  outdir             : output</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>Generic options</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>  trace_report_suffix: 2025-05-06_03-22-36</span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>Core Nextflow options</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>  runName            : goofy_cori</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>  containerEngine    : singularity</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>  launchDir          : /scratch/users/sli/workshop/pipeline</span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>  workDir            : /scratch/users/sli/workshop/pipeline/work</span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>  projectDir         : /scratch/users/sli/workshop/pipeline/nf-core-customrnaseq</span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a>  userName           : sli</span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>  profile            : singularity</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a>  configFiles        : /scratch/users/sli/workshop/pipeline/nf-core-customrnaseq/nextflow.config</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a>!! Only displaying parameters that differ from the pipeline defaults !!</span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------</span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a>* The nf-core framework</span>
<span id="cb20-35"><a href="#cb20-35" aria-hidden="true" tabindex="-1"></a>    https://doi.org/10.1038/s41587-020-0439-x</span>
<span id="cb20-36"><a href="#cb20-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-37"><a href="#cb20-37" aria-hidden="true" tabindex="-1"></a>* Software dependencies</span>
<span id="cb20-38"><a href="#cb20-38" aria-hidden="true" tabindex="-1"></a>    https://github.com/nf-core/customrnaseq/blob/main/CITATIONS.md</span>
<span id="cb20-39"><a href="#cb20-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-40"><a href="#cb20-40" aria-hidden="true" tabindex="-1"></a>WARN: The following invalid input values have been detected:</span>
<span id="cb20-41"><a href="#cb20-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-42"><a href="#cb20-42" aria-hidden="true" tabindex="-1"></a>* --transcript_fasta: /scratch/users/sli/workshop/training/nf-training/data/ggal/transcriptome.fa</span>
<span id="cb20-43"><a href="#cb20-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-44"><a href="#cb20-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-45"><a href="#cb20-45" aria-hidden="true" tabindex="-1"></a>executor &gt;  local (1)</span>
<span id="cb20-46"><a href="#cb20-46" aria-hidden="true" tabindex="-1"></a>[19/79bd6e] NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:FASTQC (liver)       | 3 of 3, cached: 3 ✔</span>
<span id="cb20-47"><a href="#cb20-47" aria-hidden="true" tabindex="-1"></a>[5c/05a99e] NFC…ASEQ:CUSTOMRNASEQ:SALMON_INDEX (transcriptome.fa) | 1 of 1 ✔</span>
<span id="cb20-48"><a href="#cb20-48" aria-hidden="true" tabindex="-1"></a>[e4/41d893] NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:MULTIQC              | 1 of 1 ✔</span>
<span id="cb20-49"><a href="#cb20-49" aria-hidden="true" tabindex="-1"></a>-[nf-core/customrnaseq] Pipeline completed successfully-</span></code></pre></div>
</div>
</div>
</div>
<p>Here, the pipeline completed successfully, with the new process <code>SALMON_INDEX</code> completing successfully. However, there is one new warning, which we will discuss more later:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>WARN: The following invalid input values have been detected:</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>* --transcript_fasta: /scratch/users/sli/workshop/training/nf-training/data/ggal/transcriptome.fa</span></code></pre></div>
</section>
<section id="process-salmonquant" class="level4">
<h4>Process <code>salmon/quant</code></h4>
<p>Let’s repeat the process for <code>salmon/quant</code>. Now that we know the contents of the parsed samplesheet channel, we need to determine if this channel is suitable to be used in our processes.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>[[name:gut, single_end:false], [/.../data/gut_1.fastq.gz, /.../data/gut_2.fastq.gz]]</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>[[name:liver, single_end:false], [/.../data/liver_1.fastq.gz, /.../data/liver_2.fastq.gz]]</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>[[name:lung, single_end:false], [/.../data/lung_1.fastq.gz, /.../data/lung_2.fastq.gz]]</span></code></pre></div>
<p>From the <code>salmon/quant</code> <a href="https://github.com/nf-core/modules/blob/master/modules/nf-core/salmon/quant/main.nf">module GitHub page</a>, we see that many inputs are needed in addition to our data.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>process SALMON_QUANT {</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    tag &quot;$meta.id&quot;</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    label &quot;process_medium&quot;</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    conda &quot;${moduleDir}/environment.yml&quot;</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    container &quot;${ workflow.containerEngine == &#39;singularity&#39; &amp;&amp; !task.ext.singularity_pull_docker_container ?</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>        &#39;https://depot.galaxyproject.org/singularity/salmon:1.10.3--h6dccd9a_2&#39; :</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>        &#39;biocontainers/salmon:1.10.3--h6dccd9a_2&#39; }&quot;</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    input:</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    tuple val(meta), path(reads)</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    path  index</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>    path  gtf</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>    path  transcript_fasta</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>    val   alignment_mode</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>    val   lib_type</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<ul>
<li>The first input that is required is a tuple of two elements – the first element containing the sample metadata, and the second element containing the paths to the FASTQ files. The default channel parsed from the samplesheet matches this structure, so it can be used as the input.</li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
tag “$meta.id”
</div>
</div>
<div class="callout-body-container callout-body">
<p>Recall that when a process is exected, the output containing the process name, along with the contents of the <code>tag</code> directive will be printed:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>[1a/915255] process &gt; FOO (omega) [100%] 3 of 3 ✔</span></code></pre></div>
<p>In the first input to the process, the sample metadata is specified – the <code>id</code> contained within this metadata is accessed (ie. the sample name), and used in the <code>tag</code> directive. This directive is typically defined as “$meta.id” for all nf-core processes, so it is recommended that your pipeline contains this <code>id</code> metadata value.</p>
</div>
</div>
<ul>
<li><p>The second input to the process is the path to the index file. The <code>SALMON_INDEX</code> output of this process can be used as input to <code>SALMON_QUANT</code></p></li>
<li><p>The third input is the GTF file. The followig file can be used:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>/.../rnaseq_data/reference/genes.gtf </span></code></pre></div></li>
<li><p>The last two inputs are values that defines how <code>SALMON_QUANT</code> will be ran. For now, let’s define these variables within the workflow block of <code>workflows/customrnaseq.nf</code>:</p></li>
</ul>
<div class="sourceCode" id="cb26"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>    def align_mode = false</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    def lib_type = &quot;A&quot;</span></code></pre></div>
<p><strong>Exercise</strong>: Following the steps we used previously to add <code>fasta</code> and <code>transcript_fasta</code> to the workflow, repeat it for <code>gtf</code>. Provide these inputs to the <code>SALMON_QUANT</code> and rerun the workflow, ensuring the new process has been executed.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li>Inside <code>nf-core-customrnaseq/main.nf</code>, the <code>gtf</code> parameter is added:</li>
</ol>
<div class="sourceCode" id="cb27"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>params.gtf = getGenomeAttribute(&#39;gtf&#39;)</span></code></pre></div>
<ol start="2" type="1">
<li>The path to the <code>gtf</code> file is specified in the <code>params.yaml</code>:</li>
</ol>
<div class="sourceCode" id="cb28"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>gtf: &quot;/.../rnaseq_data/reference/genes.gtf&quot;</span></code></pre></div>
<ol start="3" type="1">
<li>Inside <code>workflows/customrnaseq.nf</code>, convert the parameter to a channel</li>
</ol>
<div class="sourceCode" id="cb29"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>ch_gtf =  Channel.fromPath(params.gtf)</span></code></pre></div>
<ol start="4" type="1">
<li>The inputs to <code>SALMON_QUANT</code> are defined as:</li>
</ol>
<div class="sourceCode" id="cb30"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>SALMON_QUANT (</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    ch_samplesheet,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    SALMON_INDEX.out.index,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    ch_gtf,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    ch_transcript_fasta,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    align_mode,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    lib_type</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><strong>Note</strong> that the order of inputs matter, and have to match what is declared in the process input definition.</p>
<ol start="5" type="1">
<li>The new process <code>SALMON_QUANT</code> is present, when rerunning the pipeline</li>
</ol>
<div class="sourceCode" id="cb31"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>nextflow run ./nf-core-customrnaseq/main.nf -resume -profile apptainer --input ./samplesheet.csv --outdir output -params-file ./params.yaml </span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>executor &gt;  local (2)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>[8e/9633e2] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:SALMON_INDEX (transcriptome.fasta) [100%] 1 of 1, cached: 1 ✔</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>[bb/d103d3] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:SALMON_QUANT (SRR6357070)          [100%] 1 of 1 ✔</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>[fa/542072] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:FASTQC (SRR6357072)                [100%] 3 of 3, cached: 3 ✔</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>[0f/6225cc] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:MULTIQC                            [100%] 1 of 1 ✔</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>-[nf-core/customrnaseq] Pipeline completed successfully-</span></code></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
nf-core-customrnaseq/main.nf
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb33"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>#!/usr/bin/env nextflow</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>    nf-core/customrnaseq</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>    Github : https://github.com/nf-core/customrnaseq</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    Website: https://nf-co.re/customrnaseq</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>    Slack  : https://nfcore.slack.com/channels/customrnaseq</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>----------------------------------------------------------------------------------------</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>    IMPORT FUNCTIONS / MODULES / SUBWORKFLOWS / WORKFLOWS</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>include { CUSTOMRNASEQ  } from &#39;./workflows/customrnaseq&#39;</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>include { PIPELINE_INITIALISATION } from &#39;./subworkflows/local/utils_nfcore_customrnaseq_pipeline&#39;</span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>include { PIPELINE_COMPLETION     } from &#39;./subworkflows/local/utils_nfcore_customrnaseq_pipeline&#39;</span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a>include { getGenomeAttribute      } from &#39;./subworkflows/local/utils_nfcore_customrnaseq_pipeline&#39;</span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a>    GENOME PARAMETER VALUES</span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb33-28"><a href="#cb33-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-29"><a href="#cb33-29" aria-hidden="true" tabindex="-1"></a>// TODO nf-core: Remove this line if you don&#39;t need a FASTA file</span>
<span id="cb33-30"><a href="#cb33-30" aria-hidden="true" tabindex="-1"></a>//   This is an example of how to use getGenomeAttribute() to fetch parameters</span>
<span id="cb33-31"><a href="#cb33-31" aria-hidden="true" tabindex="-1"></a>//   from igenomes.config using `--genome`</span>
<span id="cb33-32"><a href="#cb33-32" aria-hidden="true" tabindex="-1"></a>params.fasta = getGenomeAttribute(&#39;fasta&#39;)</span>
<span id="cb33-33"><a href="#cb33-33" aria-hidden="true" tabindex="-1"></a>params.transcript_fasta = getGenomeAttribute(&#39;transcript_fasta&#39;)</span>
<span id="cb33-34"><a href="#cb33-34" aria-hidden="true" tabindex="-1"></a>params.gtf = getGenomeAttribute(&#39;gtf&#39;)</span>
<span id="cb33-35"><a href="#cb33-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-36"><a href="#cb33-36" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb33-37"><a href="#cb33-37" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-38"><a href="#cb33-38" aria-hidden="true" tabindex="-1"></a>    NAMED WORKFLOWS FOR PIPELINE</span>
<span id="cb33-39"><a href="#cb33-39" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-40"><a href="#cb33-40" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb33-41"><a href="#cb33-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-42"><a href="#cb33-42" aria-hidden="true" tabindex="-1"></a>//</span>
<span id="cb33-43"><a href="#cb33-43" aria-hidden="true" tabindex="-1"></a>// WORKFLOW: Run main analysis pipeline depending on type of input</span>
<span id="cb33-44"><a href="#cb33-44" aria-hidden="true" tabindex="-1"></a>//</span>
<span id="cb33-45"><a href="#cb33-45" aria-hidden="true" tabindex="-1"></a>workflow NFCORE_CUSTOMRNASEQ {</span>
<span id="cb33-46"><a href="#cb33-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-47"><a href="#cb33-47" aria-hidden="true" tabindex="-1"></a>    take:</span>
<span id="cb33-48"><a href="#cb33-48" aria-hidden="true" tabindex="-1"></a>    samplesheet // channel: samplesheet read in from --input</span>
<span id="cb33-49"><a href="#cb33-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-50"><a href="#cb33-50" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb33-51"><a href="#cb33-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-52"><a href="#cb33-52" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-53"><a href="#cb33-53" aria-hidden="true" tabindex="-1"></a>    // WORKFLOW: Run pipeline</span>
<span id="cb33-54"><a href="#cb33-54" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-55"><a href="#cb33-55" aria-hidden="true" tabindex="-1"></a>    CUSTOMRNASEQ (</span>
<span id="cb33-56"><a href="#cb33-56" aria-hidden="true" tabindex="-1"></a>        samplesheet</span>
<span id="cb33-57"><a href="#cb33-57" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb33-58"><a href="#cb33-58" aria-hidden="true" tabindex="-1"></a>    emit:</span>
<span id="cb33-59"><a href="#cb33-59" aria-hidden="true" tabindex="-1"></a>    multiqc_report = CUSTOMRNASEQ.out.multiqc_report // channel: /path/to/multiqc_report.html</span>
<span id="cb33-60"><a href="#cb33-60" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb33-61"><a href="#cb33-61" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb33-62"><a href="#cb33-62" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-63"><a href="#cb33-63" aria-hidden="true" tabindex="-1"></a>    RUN MAIN WORKFLOW</span>
<span id="cb33-64"><a href="#cb33-64" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-65"><a href="#cb33-65" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb33-66"><a href="#cb33-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-67"><a href="#cb33-67" aria-hidden="true" tabindex="-1"></a>workflow {</span>
<span id="cb33-68"><a href="#cb33-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-69"><a href="#cb33-69" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb33-70"><a href="#cb33-70" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-71"><a href="#cb33-71" aria-hidden="true" tabindex="-1"></a>    // SUBWORKFLOW: Run initialisation tasks</span>
<span id="cb33-72"><a href="#cb33-72" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-73"><a href="#cb33-73" aria-hidden="true" tabindex="-1"></a>    PIPELINE_INITIALISATION (</span>
<span id="cb33-74"><a href="#cb33-74" aria-hidden="true" tabindex="-1"></a>        params.version,</span>
<span id="cb33-75"><a href="#cb33-75" aria-hidden="true" tabindex="-1"></a>        params.validate_params,</span>
<span id="cb33-76"><a href="#cb33-76" aria-hidden="true" tabindex="-1"></a>        params.monochrome_logs,</span>
<span id="cb33-77"><a href="#cb33-77" aria-hidden="true" tabindex="-1"></a>        args,</span>
<span id="cb33-78"><a href="#cb33-78" aria-hidden="true" tabindex="-1"></a>        params.outdir,</span>
<span id="cb33-79"><a href="#cb33-79" aria-hidden="true" tabindex="-1"></a>        params.input</span>
<span id="cb33-80"><a href="#cb33-80" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb33-81"><a href="#cb33-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-82"><a href="#cb33-82" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-83"><a href="#cb33-83" aria-hidden="true" tabindex="-1"></a>    // WORKFLOW: Run main workflow</span>
<span id="cb33-84"><a href="#cb33-84" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-85"><a href="#cb33-85" aria-hidden="true" tabindex="-1"></a>    NFCORE_CUSTOMRNASEQ (</span>
<span id="cb33-86"><a href="#cb33-86" aria-hidden="true" tabindex="-1"></a>        PIPELINE_INITIALISATION.out.samplesheet</span>
<span id="cb33-87"><a href="#cb33-87" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb33-88"><a href="#cb33-88" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-89"><a href="#cb33-89" aria-hidden="true" tabindex="-1"></a>    // SUBWORKFLOW: Run completion tasks</span>
<span id="cb33-90"><a href="#cb33-90" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb33-91"><a href="#cb33-91" aria-hidden="true" tabindex="-1"></a>    PIPELINE_COMPLETION (</span>
<span id="cb33-92"><a href="#cb33-92" aria-hidden="true" tabindex="-1"></a>        params.email,</span>
<span id="cb33-93"><a href="#cb33-93" aria-hidden="true" tabindex="-1"></a>        params.email_on_fail,</span>
<span id="cb33-94"><a href="#cb33-94" aria-hidden="true" tabindex="-1"></a>        params.plaintext_email,</span>
<span id="cb33-95"><a href="#cb33-95" aria-hidden="true" tabindex="-1"></a>        params.outdir,</span>
<span id="cb33-96"><a href="#cb33-96" aria-hidden="true" tabindex="-1"></a>        params.monochrome_logs,</span>
<span id="cb33-97"><a href="#cb33-97" aria-hidden="true" tabindex="-1"></a>        params.hook_url,</span>
<span id="cb33-98"><a href="#cb33-98" aria-hidden="true" tabindex="-1"></a>        NFCORE_CUSTOMRNASEQ.out.multiqc_report</span>
<span id="cb33-99"><a href="#cb33-99" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb33-100"><a href="#cb33-100" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb33-101"><a href="#cb33-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-102"><a href="#cb33-102" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb33-103"><a href="#cb33-103" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-104"><a href="#cb33-104" aria-hidden="true" tabindex="-1"></a>    THE END</span>
<span id="cb33-105"><a href="#cb33-105" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb33-106"><a href="#cb33-106" aria-hidden="true" tabindex="-1"></a>*/</span></code></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
nf-core-customrnaseq/workflows/customrnaseq.nf
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb34"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    IMPORT MODULES / SUBWORKFLOWS / FUNCTIONS</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>include { FASTQC                 } from &#39;../modules/nf-core/fastqc/main&#39;</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>include { MULTIQC                } from &#39;../modules/nf-core/multiqc/main&#39;</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>include { paramsSummaryMap       } from &#39;plugin/nf-schema&#39;</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>include { paramsSummaryMultiqc   } from &#39;../subworkflows/nf-core/utils_nfcore_pipeline&#39;</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>include { softwareVersionsToYAML } from &#39;../subworkflows/nf-core/utils_nfcore_pipeline&#39;</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>include { methodsDescriptionText } from &#39;../subworkflows/local/utils_nfcore_customrnaseq_pipeline&#39;</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>include { SALMON_INDEX } from &#39;../modules/nf-core/salmon/index/main&#39; </span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>include { SALMON_QUANT } from &#39;../modules/nf-core/salmon/quant/main&#39;    </span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>    RUN MAIN WORKFLOW</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>*/</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>workflow CUSTOMRNASEQ {</span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>    take:</span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>    ch_samplesheet // channel: samplesheet read in from --input</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a>    main:</span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a>    // Gather reference files</span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>    ch_genome_fasta = Channel.fromPath(params.genome_fasta)</span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>    ch_transcript_fasta = Channel.fromPath(params.transcript_fasta)</span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a>    ch_gtf =  Channel.fromPath(params.gtf)</span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a>    ch_versions = Channel.empty()</span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_files = Channel.empty()</span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a>    // MODULE: Run FastQC</span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a>    FASTQC (</span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a>        ch_samplesheet</span>
<span id="cb34-42"><a href="#cb34-42" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb34-43"><a href="#cb34-43" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]})</span>
<span id="cb34-44"><a href="#cb34-44" aria-hidden="true" tabindex="-1"></a>    ch_versions = ch_versions.mix(FASTQC.out.versions.first())</span>
<span id="cb34-45"><a href="#cb34-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-46"><a href="#cb34-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-47"><a href="#cb34-47" aria-hidden="true" tabindex="-1"></a>    SALMON_INDEX ( </span>
<span id="cb34-48"><a href="#cb34-48" aria-hidden="true" tabindex="-1"></a>        ch_genome_fasta,</span>
<span id="cb34-49"><a href="#cb34-49" aria-hidden="true" tabindex="-1"></a>        ch_transcript_fasta</span>
<span id="cb34-50"><a href="#cb34-50" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb34-51"><a href="#cb34-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-52"><a href="#cb34-52" aria-hidden="true" tabindex="-1"></a>    def align_mode = false</span>
<span id="cb34-53"><a href="#cb34-53" aria-hidden="true" tabindex="-1"></a>    def lib_type = &quot;A&quot;</span>
<span id="cb34-54"><a href="#cb34-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-55"><a href="#cb34-55" aria-hidden="true" tabindex="-1"></a>    SALMON_QUANT (</span>
<span id="cb34-56"><a href="#cb34-56" aria-hidden="true" tabindex="-1"></a>        ch_samplesheet,</span>
<span id="cb34-57"><a href="#cb34-57" aria-hidden="true" tabindex="-1"></a>        SALMON_INDEX.out.index,</span>
<span id="cb34-58"><a href="#cb34-58" aria-hidden="true" tabindex="-1"></a>        ch_gtf,</span>
<span id="cb34-59"><a href="#cb34-59" aria-hidden="true" tabindex="-1"></a>        ch_transcript_fasta,</span>
<span id="cb34-60"><a href="#cb34-60" aria-hidden="true" tabindex="-1"></a>        align_mode,</span>
<span id="cb34-61"><a href="#cb34-61" aria-hidden="true" tabindex="-1"></a>        lib_type</span>
<span id="cb34-62"><a href="#cb34-62" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb34-63"><a href="#cb34-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-64"><a href="#cb34-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-65"><a href="#cb34-65" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb34-66"><a href="#cb34-66" aria-hidden="true" tabindex="-1"></a>    // Collate and save software versions</span>
<span id="cb34-67"><a href="#cb34-67" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb34-68"><a href="#cb34-68" aria-hidden="true" tabindex="-1"></a>    softwareVersionsToYAML(ch_versions)</span>
<span id="cb34-69"><a href="#cb34-69" aria-hidden="true" tabindex="-1"></a>        .collectFile(</span>
<span id="cb34-70"><a href="#cb34-70" aria-hidden="true" tabindex="-1"></a>            storeDir: &quot;${params.outdir}/pipeline_info&quot;,</span>
<span id="cb34-71"><a href="#cb34-71" aria-hidden="true" tabindex="-1"></a>            name: &#39;nf_core_&#39;  +  &#39;customrnaseq_software_&#39;  + &#39;mqc_&#39;  + &#39;versions.yml&#39;,</span>
<span id="cb34-72"><a href="#cb34-72" aria-hidden="true" tabindex="-1"></a>            sort: true,</span>
<span id="cb34-73"><a href="#cb34-73" aria-hidden="true" tabindex="-1"></a>            newLine: true</span>
<span id="cb34-74"><a href="#cb34-74" aria-hidden="true" tabindex="-1"></a>        ).set { ch_collated_versions }</span>
<span id="cb34-75"><a href="#cb34-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-76"><a href="#cb34-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-77"><a href="#cb34-77" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb34-78"><a href="#cb34-78" aria-hidden="true" tabindex="-1"></a>    // MODULE: MultiQC</span>
<span id="cb34-79"><a href="#cb34-79" aria-hidden="true" tabindex="-1"></a>    //</span>
<span id="cb34-80"><a href="#cb34-80" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_config        = Channel.fromPath(</span>
<span id="cb34-81"><a href="#cb34-81" aria-hidden="true" tabindex="-1"></a>        &quot;$projectDir/assets/multiqc_config.yml&quot;, checkIfExists: true)</span>
<span id="cb34-82"><a href="#cb34-82" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_custom_config = params.multiqc_config ?</span>
<span id="cb34-83"><a href="#cb34-83" aria-hidden="true" tabindex="-1"></a>        Channel.fromPath(params.multiqc_config, checkIfExists: true) :</span>
<span id="cb34-84"><a href="#cb34-84" aria-hidden="true" tabindex="-1"></a>        Channel.empty()</span>
<span id="cb34-85"><a href="#cb34-85" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_logo          = params.multiqc_logo ?</span>
<span id="cb34-86"><a href="#cb34-86" aria-hidden="true" tabindex="-1"></a>        Channel.fromPath(params.multiqc_logo, checkIfExists: true) :</span>
<span id="cb34-87"><a href="#cb34-87" aria-hidden="true" tabindex="-1"></a>        Channel.empty()</span>
<span id="cb34-88"><a href="#cb34-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-89"><a href="#cb34-89" aria-hidden="true" tabindex="-1"></a>    summary_params      = paramsSummaryMap(</span>
<span id="cb34-90"><a href="#cb34-90" aria-hidden="true" tabindex="-1"></a>        workflow, parameters_schema: &quot;nextflow_schema.json&quot;)</span>
<span id="cb34-91"><a href="#cb34-91" aria-hidden="true" tabindex="-1"></a>    ch_workflow_summary = Channel.value(paramsSummaryMultiqc(summary_params))</span>
<span id="cb34-92"><a href="#cb34-92" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_files = ch_multiqc_files.mix(</span>
<span id="cb34-93"><a href="#cb34-93" aria-hidden="true" tabindex="-1"></a>        ch_workflow_summary.collectFile(name: &#39;workflow_summary_mqc.yaml&#39;))</span>
<span id="cb34-94"><a href="#cb34-94" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_custom_methods_description = params.multiqc_methods_description ?</span>
<span id="cb34-95"><a href="#cb34-95" aria-hidden="true" tabindex="-1"></a>        file(params.multiqc_methods_description, checkIfExists: true) :</span>
<span id="cb34-96"><a href="#cb34-96" aria-hidden="true" tabindex="-1"></a>        file(&quot;$projectDir/assets/methods_description_template.yml&quot;, checkIfExists: true)</span>
<span id="cb34-97"><a href="#cb34-97" aria-hidden="true" tabindex="-1"></a>    ch_methods_description                = Channel.value(</span>
<span id="cb34-98"><a href="#cb34-98" aria-hidden="true" tabindex="-1"></a>        methodsDescriptionText(ch_multiqc_custom_methods_description))</span>
<span id="cb34-99"><a href="#cb34-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-100"><a href="#cb34-100" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_files = ch_multiqc_files.mix(ch_collated_versions)</span>
<span id="cb34-101"><a href="#cb34-101" aria-hidden="true" tabindex="-1"></a>    ch_multiqc_files = ch_multiqc_files.mix(</span>
<span id="cb34-102"><a href="#cb34-102" aria-hidden="true" tabindex="-1"></a>        ch_methods_description.collectFile(</span>
<span id="cb34-103"><a href="#cb34-103" aria-hidden="true" tabindex="-1"></a>            name: &#39;methods_description_mqc.yaml&#39;,</span>
<span id="cb34-104"><a href="#cb34-104" aria-hidden="true" tabindex="-1"></a>            sort: true</span>
<span id="cb34-105"><a href="#cb34-105" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb34-106"><a href="#cb34-106" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb34-107"><a href="#cb34-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-108"><a href="#cb34-108" aria-hidden="true" tabindex="-1"></a>    MULTIQC (</span>
<span id="cb34-109"><a href="#cb34-109" aria-hidden="true" tabindex="-1"></a>        ch_multiqc_files.collect(),</span>
<span id="cb34-110"><a href="#cb34-110" aria-hidden="true" tabindex="-1"></a>        ch_multiqc_config.toList(),</span>
<span id="cb34-111"><a href="#cb34-111" aria-hidden="true" tabindex="-1"></a>        ch_multiqc_custom_config.toList(),</span>
<span id="cb34-112"><a href="#cb34-112" aria-hidden="true" tabindex="-1"></a>        ch_multiqc_logo.toList(),</span>
<span id="cb34-113"><a href="#cb34-113" aria-hidden="true" tabindex="-1"></a>        [],</span>
<span id="cb34-114"><a href="#cb34-114" aria-hidden="true" tabindex="-1"></a>        []</span>
<span id="cb34-115"><a href="#cb34-115" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb34-116"><a href="#cb34-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-117"><a href="#cb34-117" aria-hidden="true" tabindex="-1"></a>    emit:multiqc_report = MULTIQC.out.report.toList() // channel: /path/to/multiqc_report.html</span>
<span id="cb34-118"><a href="#cb34-118" aria-hidden="true" tabindex="-1"></a>    versions       = ch_versions                 // channel: [ path(versions.yml) ]</span>
<span id="cb34-119"><a href="#cb34-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-120"><a href="#cb34-120" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-121"><a href="#cb34-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-122"><a href="#cb34-122" aria-hidden="true" tabindex="-1"></a>/*</span>
<span id="cb34-123"><a href="#cb34-123" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb34-124"><a href="#cb34-124" aria-hidden="true" tabindex="-1"></a>    THE END</span>
<span id="cb34-125"><a href="#cb34-125" aria-hidden="true" tabindex="-1"></a>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>
<span id="cb34-126"><a href="#cb34-126" aria-hidden="true" tabindex="-1"></a>*/</span></code></pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="schema-validation" class="level2">
<h2><strong>6.2 Schema validation</strong></h2>
<p>When running the pipeline you may have noticed that although the pipeline completed successfully, there were warnings output:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>WARN: The following invalid input values have been detected:</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>* --transcript_fasta: /home/goldmire/rnaseq_data/reference/transcriptome.fasta</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>* --gtf: /home/goldmire/rnaseq_data/reference/genes.gtf</span></code></pre></div>
<p>These warnings are because the new input parameters we specified to the pipeline were not included in the <code>nf-core-customrnaseq/nextflow_schema.json</code> file.</p>
<p>This file contains all the information about pipeline configuration parameters.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
<code>nextflow_schema.json</code>
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb36"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>    &quot;$schema&quot;: &quot;https://json-schema.org/draft/2020-12/schema&quot;,</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    &quot;$id&quot;: &quot;https://raw.githubusercontent.com/nf-core/customrnaseq/master/nextflow_schema.json&quot;,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>    &quot;title&quot;: &quot;nf-core/customrnaseq pipeline parameters&quot;,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    &quot;description&quot;: &quot;Custom RNASeq pipeline&quot;,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>    &quot;type&quot;: &quot;object&quot;,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>    &quot;$defs&quot;: {</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>        &quot;input_output_options&quot;: {</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>            &quot;title&quot;: &quot;Input/output options&quot;,</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>            &quot;type&quot;: &quot;object&quot;,</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>            &quot;fa_icon&quot;: &quot;fas fa-terminal&quot;,</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>            &quot;description&quot;: &quot;Define where the pipeline should find input data and save output data.&quot;,</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>            &quot;required&quot;: [&quot;input&quot;, &quot;outdir&quot;],</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>            &quot;properties&quot;: {</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>                &quot;input&quot;: {</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>                    &quot;format&quot;: &quot;file-path&quot;,</span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>                    &quot;exists&quot;: true,</span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a>                    &quot;schema&quot;: &quot;assets/schema_input.json&quot;,</span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>                    &quot;mimetype&quot;: &quot;text/csv&quot;,</span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>                    &quot;pattern&quot;: &quot;^\\S+\\.csv$&quot;,</span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Path to comma-separated file containing information about the samples in the experiment.&quot;,</span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;You will need to create a design file with information about the samples in your experiment before running the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 3 columns, and a header row. See [usage docs](https://nf-co.re/customrnaseq/usage#samplesheet-input).&quot;,</span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-file-csv&quot;</span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>                &quot;outdir&quot;: {</span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a>                    &quot;format&quot;: &quot;directory-path&quot;,</span>
<span id="cb36-29"><a href="#cb36-29" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;The output directory where the results will be saved. You have to use absolute paths to storage on Cloud infrastructure.&quot;,</span>
<span id="cb36-30"><a href="#cb36-30" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-folder-open&quot;</span>
<span id="cb36-31"><a href="#cb36-31" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-32"><a href="#cb36-32" aria-hidden="true" tabindex="-1"></a>                &quot;email&quot;: {</span>
<span id="cb36-33"><a href="#cb36-33" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-34"><a href="#cb36-34" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Email address for completion summary.&quot;,</span>
<span id="cb36-35"><a href="#cb36-35" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-envelope&quot;,</span>
<span id="cb36-36"><a href="#cb36-36" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits. If set in your user config file (`~/.nextflow/config`) then you don&#39;t need to specify this on the command line for every run.&quot;,</span>
<span id="cb36-37"><a href="#cb36-37" aria-hidden="true" tabindex="-1"></a>                    &quot;pattern&quot;: &quot;^([a-zA-Z0-9_\\-\\.]+)@([a-zA-Z0-9_\\-\\.]+)\\.([a-zA-Z]{2,5})$&quot;</span>
<span id="cb36-38"><a href="#cb36-38" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-39"><a href="#cb36-39" aria-hidden="true" tabindex="-1"></a>                &quot;multiqc_title&quot;: {</span>
<span id="cb36-40"><a href="#cb36-40" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-41"><a href="#cb36-41" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;MultiQC report title. Printed as page header, used for filename if not otherwise specified.&quot;,</span>
<span id="cb36-42"><a href="#cb36-42" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-file-signature&quot;</span>
<span id="cb36-43"><a href="#cb36-43" aria-hidden="true" tabindex="-1"></a>                }</span>
<span id="cb36-44"><a href="#cb36-44" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb36-45"><a href="#cb36-45" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb36-46"><a href="#cb36-46" aria-hidden="true" tabindex="-1"></a>        &quot;reference_genome_options&quot;: {</span>
<span id="cb36-47"><a href="#cb36-47" aria-hidden="true" tabindex="-1"></a>            &quot;title&quot;: &quot;Reference genome options&quot;,</span>
<span id="cb36-48"><a href="#cb36-48" aria-hidden="true" tabindex="-1"></a>            &quot;type&quot;: &quot;object&quot;,</span>
<span id="cb36-49"><a href="#cb36-49" aria-hidden="true" tabindex="-1"></a>            &quot;fa_icon&quot;: &quot;fas fa-dna&quot;,</span>
<span id="cb36-50"><a href="#cb36-50" aria-hidden="true" tabindex="-1"></a>            &quot;description&quot;: &quot;Reference genome related files and options required for the workflow.&quot;,</span>
<span id="cb36-51"><a href="#cb36-51" aria-hidden="true" tabindex="-1"></a>            &quot;properties&quot;: {</span>
<span id="cb36-52"><a href="#cb36-52" aria-hidden="true" tabindex="-1"></a>                &quot;genome&quot;: {</span>
<span id="cb36-53"><a href="#cb36-53" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-54"><a href="#cb36-54" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Name of iGenomes reference.&quot;,</span>
<span id="cb36-55"><a href="#cb36-55" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-book&quot;,</span>
<span id="cb36-56"><a href="#cb36-56" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;If using a reference genome configured in the pipeline using iGenomes, use this parameter to give the ID for the reference. This is then used to build the full paths for all required reference genome files e.g. `--genome GRCh38`. \n\nSee the [nf-core website docs](https://nf-co.re/usage/reference_genomes) for more details.&quot;</span>
<span id="cb36-57"><a href="#cb36-57" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-58"><a href="#cb36-58" aria-hidden="true" tabindex="-1"></a>                &quot;fasta&quot;: {</span>
<span id="cb36-59"><a href="#cb36-59" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-60"><a href="#cb36-60" aria-hidden="true" tabindex="-1"></a>                    &quot;format&quot;: &quot;file-path&quot;,</span>
<span id="cb36-61"><a href="#cb36-61" aria-hidden="true" tabindex="-1"></a>                    &quot;exists&quot;: true,</span>
<span id="cb36-62"><a href="#cb36-62" aria-hidden="true" tabindex="-1"></a>                    &quot;mimetype&quot;: &quot;text/plain&quot;,</span>
<span id="cb36-63"><a href="#cb36-63" aria-hidden="true" tabindex="-1"></a>                    &quot;pattern&quot;: &quot;^\\S+\\.fn?a(sta)?(\\.gz)?$&quot;,</span>
<span id="cb36-64"><a href="#cb36-64" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Path to FASTA genome file.&quot;,</span>
<span id="cb36-65"><a href="#cb36-65" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;This parameter is *mandatory* if `--genome` is not specified. If you don&#39;t have a BWA index available this will be generated for you automatically. Combine with `--save_reference` to save BWA index for future runs.&quot;,</span>
<span id="cb36-66"><a href="#cb36-66" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;far fa-file-code&quot;</span>
<span id="cb36-67"><a href="#cb36-67" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-68"><a href="#cb36-68" aria-hidden="true" tabindex="-1"></a>                &quot;igenomes_ignore&quot;: {</span>
<span id="cb36-69"><a href="#cb36-69" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;boolean&quot;,</span>
<span id="cb36-70"><a href="#cb36-70" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Do not load the iGenomes reference config.&quot;,</span>
<span id="cb36-71"><a href="#cb36-71" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-ban&quot;,</span>
<span id="cb36-72"><a href="#cb36-72" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-73"><a href="#cb36-73" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;Do not load `igenomes.config` when running the pipeline. You may choose this option if you observe clashes between custom parameters and those supplied in `igenomes.config`.&quot;</span>
<span id="cb36-74"><a href="#cb36-74" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-75"><a href="#cb36-75" aria-hidden="true" tabindex="-1"></a>                &quot;igenomes_base&quot;: {</span>
<span id="cb36-76"><a href="#cb36-76" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-77"><a href="#cb36-77" aria-hidden="true" tabindex="-1"></a>                    &quot;format&quot;: &quot;directory-path&quot;,</span>
<span id="cb36-78"><a href="#cb36-78" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;The base path to the igenomes reference files&quot;,</span>
<span id="cb36-79"><a href="#cb36-79" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-ban&quot;,</span>
<span id="cb36-80"><a href="#cb36-80" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-81"><a href="#cb36-81" aria-hidden="true" tabindex="-1"></a>                    &quot;default&quot;: &quot;s3://ngi-igenomes/igenomes/&quot;</span>
<span id="cb36-82"><a href="#cb36-82" aria-hidden="true" tabindex="-1"></a>                }</span>
<span id="cb36-83"><a href="#cb36-83" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb36-84"><a href="#cb36-84" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb36-85"><a href="#cb36-85" aria-hidden="true" tabindex="-1"></a>        &quot;institutional_config_options&quot;: {</span>
<span id="cb36-86"><a href="#cb36-86" aria-hidden="true" tabindex="-1"></a>            &quot;title&quot;: &quot;Institutional config options&quot;,</span>
<span id="cb36-87"><a href="#cb36-87" aria-hidden="true" tabindex="-1"></a>            &quot;type&quot;: &quot;object&quot;,</span>
<span id="cb36-88"><a href="#cb36-88" aria-hidden="true" tabindex="-1"></a>            &quot;fa_icon&quot;: &quot;fas fa-university&quot;,</span>
<span id="cb36-89"><a href="#cb36-89" aria-hidden="true" tabindex="-1"></a>            &quot;description&quot;: &quot;Parameters used to describe centralised config profiles. These should not be edited.&quot;,</span>
<span id="cb36-90"><a href="#cb36-90" aria-hidden="true" tabindex="-1"></a>            &quot;help_text&quot;: &quot;The centralised nf-core configuration profiles use a handful of pipeline parameters to describe themselves. This information is then printed to the Nextflow log when you run a pipeline. You should not need to change these values when you run a pipeline.&quot;,</span>
<span id="cb36-91"><a href="#cb36-91" aria-hidden="true" tabindex="-1"></a>            &quot;properties&quot;: {</span>
<span id="cb36-92"><a href="#cb36-92" aria-hidden="true" tabindex="-1"></a>                &quot;custom_config_version&quot;: {</span>
<span id="cb36-93"><a href="#cb36-93" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-94"><a href="#cb36-94" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Git commit id for Institutional configs.&quot;,</span>
<span id="cb36-95"><a href="#cb36-95" aria-hidden="true" tabindex="-1"></a>                    &quot;default&quot;: &quot;master&quot;,</span>
<span id="cb36-96"><a href="#cb36-96" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-97"><a href="#cb36-97" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-users-cog&quot;</span>
<span id="cb36-98"><a href="#cb36-98" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-99"><a href="#cb36-99" aria-hidden="true" tabindex="-1"></a>                &quot;custom_config_base&quot;: {</span>
<span id="cb36-100"><a href="#cb36-100" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-101"><a href="#cb36-101" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Base directory for Institutional configs.&quot;,</span>
<span id="cb36-102"><a href="#cb36-102" aria-hidden="true" tabindex="-1"></a>                    &quot;default&quot;: &quot;https://raw.githubusercontent.com/nf-core/configs/master&quot;,</span>
<span id="cb36-103"><a href="#cb36-103" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-104"><a href="#cb36-104" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;If you&#39;re running offline, Nextflow will not be able to fetch the institutional config files from the internet. If you don&#39;t need them, then this is not a problem. If you do need them, you should download the files from the repo and tell Nextflow where to find them with this parameter.&quot;,</span>
<span id="cb36-105"><a href="#cb36-105" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-users-cog&quot;</span>
<span id="cb36-106"><a href="#cb36-106" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-107"><a href="#cb36-107" aria-hidden="true" tabindex="-1"></a>                &quot;config_profile_name&quot;: {</span>
<span id="cb36-108"><a href="#cb36-108" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-109"><a href="#cb36-109" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Institutional config name.&quot;,</span>
<span id="cb36-110"><a href="#cb36-110" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-111"><a href="#cb36-111" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-users-cog&quot;</span>
<span id="cb36-112"><a href="#cb36-112" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-113"><a href="#cb36-113" aria-hidden="true" tabindex="-1"></a>                &quot;config_profile_description&quot;: {</span>
<span id="cb36-114"><a href="#cb36-114" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-115"><a href="#cb36-115" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Institutional config description.&quot;,</span>
<span id="cb36-116"><a href="#cb36-116" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-117"><a href="#cb36-117" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-users-cog&quot;</span>
<span id="cb36-118"><a href="#cb36-118" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-119"><a href="#cb36-119" aria-hidden="true" tabindex="-1"></a>                &quot;config_profile_contact&quot;: {</span>
<span id="cb36-120"><a href="#cb36-120" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-121"><a href="#cb36-121" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Institutional config contact information.&quot;,</span>
<span id="cb36-122"><a href="#cb36-122" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-123"><a href="#cb36-123" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-users-cog&quot;</span>
<span id="cb36-124"><a href="#cb36-124" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-125"><a href="#cb36-125" aria-hidden="true" tabindex="-1"></a>                &quot;config_profile_url&quot;: {</span>
<span id="cb36-126"><a href="#cb36-126" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-127"><a href="#cb36-127" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Institutional config URL link.&quot;,</span>
<span id="cb36-128"><a href="#cb36-128" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true,</span>
<span id="cb36-129"><a href="#cb36-129" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-users-cog&quot;</span>
<span id="cb36-130"><a href="#cb36-130" aria-hidden="true" tabindex="-1"></a>                }</span>
<span id="cb36-131"><a href="#cb36-131" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb36-132"><a href="#cb36-132" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb36-133"><a href="#cb36-133" aria-hidden="true" tabindex="-1"></a>        &quot;generic_options&quot;: {</span>
<span id="cb36-134"><a href="#cb36-134" aria-hidden="true" tabindex="-1"></a>            &quot;title&quot;: &quot;Generic options&quot;,</span>
<span id="cb36-135"><a href="#cb36-135" aria-hidden="true" tabindex="-1"></a>            &quot;type&quot;: &quot;object&quot;,</span>
<span id="cb36-136"><a href="#cb36-136" aria-hidden="true" tabindex="-1"></a>            &quot;fa_icon&quot;: &quot;fas fa-file-import&quot;,</span>
<span id="cb36-137"><a href="#cb36-137" aria-hidden="true" tabindex="-1"></a>            &quot;description&quot;: &quot;Less common options for the pipeline, typically set in a config file.&quot;,</span>
<span id="cb36-138"><a href="#cb36-138" aria-hidden="true" tabindex="-1"></a>            &quot;help_text&quot;: &quot;These options are common to all nf-core pipelines and allow you to customise some of the core preferences for how the pipeline runs.\n\nTypically these options would be set in a Nextflow config file loaded for all pipeline runs, such as `~/.nextflow/config`.&quot;,</span>
<span id="cb36-139"><a href="#cb36-139" aria-hidden="true" tabindex="-1"></a>            &quot;properties&quot;: {</span>
<span id="cb36-140"><a href="#cb36-140" aria-hidden="true" tabindex="-1"></a>                &quot;version&quot;: {</span>
<span id="cb36-141"><a href="#cb36-141" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;boolean&quot;,</span>
<span id="cb36-142"><a href="#cb36-142" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Display version and exit.&quot;,</span>
<span id="cb36-143"><a href="#cb36-143" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-question-circle&quot;,</span>
<span id="cb36-144"><a href="#cb36-144" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-145"><a href="#cb36-145" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-146"><a href="#cb36-146" aria-hidden="true" tabindex="-1"></a>                &quot;publish_dir_mode&quot;: {</span>
<span id="cb36-147"><a href="#cb36-147" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-148"><a href="#cb36-148" aria-hidden="true" tabindex="-1"></a>                    &quot;default&quot;: &quot;copy&quot;,</span>
<span id="cb36-149"><a href="#cb36-149" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Method used to save pipeline results to output directory.&quot;,</span>
<span id="cb36-150"><a href="#cb36-150" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;The Nextflow `publishDir` option specifies which intermediate files should be saved to the output directory. This option tells the pipeline what method should be used to move these files. See [Nextflow docs](https://www.nextflow.io/docs/latest/process.html#publishdir) for details.&quot;,</span>
<span id="cb36-151"><a href="#cb36-151" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-copy&quot;,</span>
<span id="cb36-152"><a href="#cb36-152" aria-hidden="true" tabindex="-1"></a>                    &quot;enum&quot;: [&quot;symlink&quot;, &quot;rellink&quot;, &quot;link&quot;, &quot;copy&quot;, &quot;copyNoFollow&quot;, &quot;move&quot;],</span>
<span id="cb36-153"><a href="#cb36-153" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-154"><a href="#cb36-154" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-155"><a href="#cb36-155" aria-hidden="true" tabindex="-1"></a>                &quot;email_on_fail&quot;: {</span>
<span id="cb36-156"><a href="#cb36-156" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-157"><a href="#cb36-157" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Email address for completion summary, only when pipeline fails.&quot;,</span>
<span id="cb36-158"><a href="#cb36-158" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-exclamation-triangle&quot;,</span>
<span id="cb36-159"><a href="#cb36-159" aria-hidden="true" tabindex="-1"></a>                    &quot;pattern&quot;: &quot;^([a-zA-Z0-9_\\-\\.]+)@([a-zA-Z0-9_\\-\\.]+)\\.([a-zA-Z]{2,5})$&quot;,</span>
<span id="cb36-160"><a href="#cb36-160" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;An email address to send a summary email to when the pipeline is completed - ONLY sent if the pipeline does not exit successfully.&quot;,</span>
<span id="cb36-161"><a href="#cb36-161" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-162"><a href="#cb36-162" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-163"><a href="#cb36-163" aria-hidden="true" tabindex="-1"></a>                &quot;plaintext_email&quot;: {</span>
<span id="cb36-164"><a href="#cb36-164" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;boolean&quot;,</span>
<span id="cb36-165"><a href="#cb36-165" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Send plain-text email instead of HTML.&quot;,</span>
<span id="cb36-166"><a href="#cb36-166" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-remove-format&quot;,</span>
<span id="cb36-167"><a href="#cb36-167" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-168"><a href="#cb36-168" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-169"><a href="#cb36-169" aria-hidden="true" tabindex="-1"></a>                &quot;max_multiqc_email_size&quot;: {</span>
<span id="cb36-170"><a href="#cb36-170" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-171"><a href="#cb36-171" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;File size limit when attaching MultiQC reports to summary emails.&quot;,</span>
<span id="cb36-172"><a href="#cb36-172" aria-hidden="true" tabindex="-1"></a>                    &quot;pattern&quot;: &quot;^\\d+(\\.\\d+)?\\.?\\s*(K|M|G|T)?B$&quot;,</span>
<span id="cb36-173"><a href="#cb36-173" aria-hidden="true" tabindex="-1"></a>                    &quot;default&quot;: &quot;25.MB&quot;,</span>
<span id="cb36-174"><a href="#cb36-174" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-file-upload&quot;,</span>
<span id="cb36-175"><a href="#cb36-175" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-176"><a href="#cb36-176" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-177"><a href="#cb36-177" aria-hidden="true" tabindex="-1"></a>                &quot;monochrome_logs&quot;: {</span>
<span id="cb36-178"><a href="#cb36-178" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;boolean&quot;,</span>
<span id="cb36-179"><a href="#cb36-179" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Do not use coloured log outputs.&quot;,</span>
<span id="cb36-180"><a href="#cb36-180" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-palette&quot;,</span>
<span id="cb36-181"><a href="#cb36-181" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-182"><a href="#cb36-182" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-183"><a href="#cb36-183" aria-hidden="true" tabindex="-1"></a>                &quot;hook_url&quot;: {</span>
<span id="cb36-184"><a href="#cb36-184" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-185"><a href="#cb36-185" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Incoming hook URL for messaging service&quot;,</span>
<span id="cb36-186"><a href="#cb36-186" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-people-group&quot;,</span>
<span id="cb36-187"><a href="#cb36-187" aria-hidden="true" tabindex="-1"></a>                    &quot;help_text&quot;: &quot;Incoming hook URL for messaging service. Currently, MS Teams and Slack are supported.&quot;,</span>
<span id="cb36-188"><a href="#cb36-188" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-189"><a href="#cb36-189" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-190"><a href="#cb36-190" aria-hidden="true" tabindex="-1"></a>                &quot;multiqc_config&quot;: {</span>
<span id="cb36-191"><a href="#cb36-191" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-192"><a href="#cb36-192" aria-hidden="true" tabindex="-1"></a>                    &quot;format&quot;: &quot;file-path&quot;,</span>
<span id="cb36-193"><a href="#cb36-193" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Custom config file to supply to MultiQC.&quot;,</span>
<span id="cb36-194"><a href="#cb36-194" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-cog&quot;,</span>
<span id="cb36-195"><a href="#cb36-195" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-196"><a href="#cb36-196" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-197"><a href="#cb36-197" aria-hidden="true" tabindex="-1"></a>                &quot;multiqc_logo&quot;: {</span>
<span id="cb36-198"><a href="#cb36-198" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-199"><a href="#cb36-199" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Custom logo file to supply to MultiQC. File name must also be set in the MultiQC config file&quot;,</span>
<span id="cb36-200"><a href="#cb36-200" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-image&quot;,</span>
<span id="cb36-201"><a href="#cb36-201" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-202"><a href="#cb36-202" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-203"><a href="#cb36-203" aria-hidden="true" tabindex="-1"></a>                &quot;multiqc_methods_description&quot;: {</span>
<span id="cb36-204"><a href="#cb36-204" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-205"><a href="#cb36-205" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Custom MultiQC yaml file containing HTML including a methods description.&quot;,</span>
<span id="cb36-206"><a href="#cb36-206" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-cog&quot;</span>
<span id="cb36-207"><a href="#cb36-207" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-208"><a href="#cb36-208" aria-hidden="true" tabindex="-1"></a>                &quot;validate_params&quot;: {</span>
<span id="cb36-209"><a href="#cb36-209" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;boolean&quot;,</span>
<span id="cb36-210"><a href="#cb36-210" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Boolean whether to validate parameters against the schema at runtime&quot;,</span>
<span id="cb36-211"><a href="#cb36-211" aria-hidden="true" tabindex="-1"></a>                    &quot;default&quot;: true,</span>
<span id="cb36-212"><a href="#cb36-212" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;fas fa-check-square&quot;,</span>
<span id="cb36-213"><a href="#cb36-213" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-214"><a href="#cb36-214" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-215"><a href="#cb36-215" aria-hidden="true" tabindex="-1"></a>                &quot;pipelines_testdata_base_path&quot;: {</span>
<span id="cb36-216"><a href="#cb36-216" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-217"><a href="#cb36-217" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;far fa-check-circle&quot;,</span>
<span id="cb36-218"><a href="#cb36-218" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Base URL or local path to location of pipeline test dataset files&quot;,</span>
<span id="cb36-219"><a href="#cb36-219" aria-hidden="true" tabindex="-1"></a>                    &quot;default&quot;: &quot;https://raw.githubusercontent.com/nf-core/test-datasets/&quot;,</span>
<span id="cb36-220"><a href="#cb36-220" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-221"><a href="#cb36-221" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb36-222"><a href="#cb36-222" aria-hidden="true" tabindex="-1"></a>                &quot;trace_report_suffix&quot;: {</span>
<span id="cb36-223"><a href="#cb36-223" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb36-224"><a href="#cb36-224" aria-hidden="true" tabindex="-1"></a>                    &quot;fa_icon&quot;: &quot;far calendar&quot;,</span>
<span id="cb36-225"><a href="#cb36-225" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Suffix to add to the trace report filename. Default is the date and time in the format yyyy-MM-dd_HH-mm-ss.&quot;,</span>
<span id="cb36-226"><a href="#cb36-226" aria-hidden="true" tabindex="-1"></a>                    &quot;hidden&quot;: true</span>
<span id="cb36-227"><a href="#cb36-227" aria-hidden="true" tabindex="-1"></a>                }</span>
<span id="cb36-228"><a href="#cb36-228" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb36-229"><a href="#cb36-229" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb36-230"><a href="#cb36-230" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="cb36-231"><a href="#cb36-231" aria-hidden="true" tabindex="-1"></a>    &quot;allOf&quot;: [</span>
<span id="cb36-232"><a href="#cb36-232" aria-hidden="true" tabindex="-1"></a>        {</span>
<span id="cb36-233"><a href="#cb36-233" aria-hidden="true" tabindex="-1"></a>            &quot;$ref&quot;: &quot;#/$defs/input_output_options&quot;</span>
<span id="cb36-234"><a href="#cb36-234" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb36-235"><a href="#cb36-235" aria-hidden="true" tabindex="-1"></a>        {</span>
<span id="cb36-236"><a href="#cb36-236" aria-hidden="true" tabindex="-1"></a>            &quot;$ref&quot;: &quot;#/$defs/reference_genome_options&quot;</span>
<span id="cb36-237"><a href="#cb36-237" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb36-238"><a href="#cb36-238" aria-hidden="true" tabindex="-1"></a>        {</span>
<span id="cb36-239"><a href="#cb36-239" aria-hidden="true" tabindex="-1"></a>            &quot;$ref&quot;: &quot;#/$defs/institutional_config_options&quot;</span>
<span id="cb36-240"><a href="#cb36-240" aria-hidden="true" tabindex="-1"></a>        },</span>
<span id="cb36-241"><a href="#cb36-241" aria-hidden="true" tabindex="-1"></a>        {</span>
<span id="cb36-242"><a href="#cb36-242" aria-hidden="true" tabindex="-1"></a>            &quot;$ref&quot;: &quot;#/$defs/generic_options&quot;</span>
<span id="cb36-243"><a href="#cb36-243" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb36-244"><a href="#cb36-244" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb36-245"><a href="#cb36-245" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
</div>
</div>
<p>Inside this file, the main parameters are separated into groups:</p>
<ul>
<li><code>input_output_options</code>: defines input and output parameter options to the pipeline</li>
<li><code>reference_genome_options</code>: defines reference parameters in the pipeline. Notice that <code>fasta</code> has already been included by default, so no warning about that parameter was output by the pipeline</li>
<li><code>institutional_config_options</code>: defines institution parameters such as profile names and custon institutional config versions</li>
<li><code>generic_options</code>: defines pipeline versions, email mode, additional configs, etc.</li>
</ul>
<p>To edit this <code>nextflow_schema.config</code> file, the following nf-core command can be used <strong>inside the pipeline folder</strong>:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>cd ./nf-core-customrnaseq</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>nf-core pipelines schema build</span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>                                          ,--./,-.</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>          ___     __   __   __   ___     /,-._.--~\ </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    |\ | |__  __ /  ` /  \ |__) |__         }  {</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    | \| |       \__, \__/ |  \ |___     \`-._,-`-,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>                                          `._,._,&#39;</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    nf-core/tools version 3.2.0 - https://nf-co.re</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    There is a new version of nf-core/tools available! (3.2.1)</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>INFO     [✓] Default parameters match schema validation                                                                                                                                                                                   </span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>INFO     [✓] Pipeline schema looks valid (found 27 params)                                                                                                                                                                                </span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>✨ Found &#39;params.transcript_fasta&#39; in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]: y</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>✨ Found &#39;params.gtf&#39; in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]: y</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>INFO     Writing schema with 29 params: &#39;nextflow_schema.json&#39;                                                                                                                                                                            </span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>🚀  Launch web builder for customisation and editing? [y/n]:  y</span></code></pre></div>
<p>This command will search for any parameters that are not defined in the schema, and ask if you would like to add them to the file. When editing, it is recommended to launch the web builder.</p>
<p>The new parameters have been added to the bottom of the parameter list:</p>
<p><img src="media/5.3_schema.png" width="700"></p>
<p>You have the option to add a parameter description, change the parameter type, specify a default value, and set if the parameter is required. Then, click and drag to the relevant section in the schema. We will drag this into the <code>reference_genome_options</code> section.</p>
<p>Once done, click <code>Finished</code> in the top right corner, and exit the window – the <code>nextflow_schema.json</code> file will be automatically updated. Open the file and check that the two new parameters have been added:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>                &quot;transcript_fasta&quot;: {</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;Transcript FASTA file&quot;</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>                },</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>                &quot;gtf&quot;: {</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>                    &quot;type&quot;: &quot;string&quot;,</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>                    &quot;description&quot;: &quot;GTF file&quot;</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>                }</span></code></pre></div>
<p>Rerun the pipeline, and check that there are no input parameter warning messages:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>nextflow run ./nf-core-customrnaseq/main.nf -resume -profile apptainer --input ./samplesheet.csv --outdir output -params-file ./params.yaml </span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="sourceCode default"><code class="sourceCode default"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>executor &gt;  local (1)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>[8e/9633e2] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:SALMON_INDEX (transcriptome.fasta) [100%] 1 of 1, cached: 1 ✔</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>[bb/d103d3] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:SALMON_QUANT (SRR6357070)          [100%] 1 of 1, cached: 1 ✔</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>[fa/542072] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:FASTQC (SRR6357072)                [100%] 3 of 3, cached: 3 ✔</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>[d4/c1e1f9] process &gt; NFCORE_CUSTOMRNASEQ:CUSTOMRNASEQ:MULTIQC                            [100%] 1 of 1 ✔</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>-[nf-core/customrnaseq] Pipeline completed successfully-</span></code></pre></div>
<div id="quarto-navigation-envelope" class="hidden">
<p><span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1zaWRlYmFyLXRpdGxl">Peter Mac Nextflow Workshop</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXItdGl0bGU=">Peter Mac Nextflow Workshop</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6SG9tZQ==">Home</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L2luZGV4Lmh0bWw=">/index.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6U2V0dXA=">Setup</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8wMF9zZXR1cC5odG1s">/workshops/00_setup.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6U2Vzc2lvbnM=">Sessions</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6RGF5IDE6IEludHJvZHVjdGlvbiB0byBOZXh0ZmxvdyBhbmQgTmYtY29yZQ==">Day 1: Introduction to Nextflow and Nf-core</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3Nlc3Npb25zLzFfaW50cm9fcnVuX25mLmh0bWw=">/sessions/1_intro_run_nf.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6RGF5IDI6IERldmVsb3BpbmcgYmlvaW5mb3JtYXRpY3Mgd29ya2Zsb3dzIHdpdGggTmV4dGZsb3c=">Day 2: Developing bioinformatics workflows with Nextflow</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3Nlc3Npb25zLzJfbmZfZGV2X2ludHJvLmh0bWw=">/sessions/2_nf_dev_intro.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6V29ya3Nob3A=">Workshop</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6MS4xIEludHJvZHVjdGlvbiB0byBOZXh0Zmxvdw==">1.1 Introduction to Nextflow</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8xLjFfaW50cm9fbmV4dGZsb3cuaHRtbA==">/workshops/1.1_intro_nextflow.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6MS4yIEludHJvZHVjdGlvbiB0byBuZi1jb3Jl">1.2 Introduction to nf-core</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8xLjJfaW50cm9fbmZfY29yZS5odG1s">/workshops/1.2_intro_nf_core.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6Mi4xIEN1c3RvbWlzaW5nICYgcnVubmluZyBuZi1jb3JlIHBpcGVsaW5lcw==">2.1 Customising &amp; running nf-core pipelines</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8yLjFfY3VzdG9taXNlX2FuZF9ydW4uaHRtbA==">/workshops/2.1_customise_and_run.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6Mi4yIFRyb3VibGVzaG9vdGluZyBOZXh0ZmxvdyBydW4=">2.2 Troubleshooting Nextflow run</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8yLjJfdHJvdWJsZXNob290aW5nLmh0bWw=">/workshops/2.2_troubleshooting.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6My4xIEludHJvZHVjdGlvbiB0byBwcm9jZXNzZXMgYW5kIGNoYW5uZWxz">3.1 Introduction to processes and channels</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8xLjNfcHJvY2Vzc2VzX2FuZF9jaGFubmVscy5odG1s">/workshops/1.3_processes_and_channels.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6My4yIENyZWF0aW5nIGEgd29ya2Zsb3c=">3.2 Creating a workflow</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8zLjFfY3JlYXRpbmdfYV93b3JrZmxvdy5odG1s">/workshops/3.1_creating_a_workflow.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6My4zIEJlc3QgcHJhY3Rpc2UsIHRpcHMgYW5kIHRyaWNrcw==">3.3 Best practise, tips and tricks</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy8yLjNfdGlwc19hbmRfdHJpY2tzLmh0bWw=">/workshops/2.3_tips_and_tricks.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6NC4xIENyZWF0aW5nIHdvcmtmbG93cyB1c2luZyBuZi1jb3JlIHRlbXBsYXRlcw==">4.1 Creating workflows using nf-core templates</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy81LjFfbmZfY29yZV90ZW1wbGF0ZS5odG1s">/workshops/5.1_nf_core_template.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6NS4xIG5mLWNvcmUgbW9kdWxlcyBhbmQgc3Vid29ya2Zsb3dz">5.1 nf-core modules and subworkflows</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy81LjJfbmZfY29yZV9tb2R1bGVzLmh0bWw=">/workshops/5.2_nf_core_modules.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6Ni4xIFNhbXBsZXNoZWV0IHBhcnNpbmcgYW5kIG5mLWNvcmUgbW9kdWxlcw==">6.1 Samplesheet parsing and nf-core modules</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy81LjNfc2FtcGxlc2hlZXRfcGFyc2luZy5odG1s">/workshops/5.3_samplesheet_parsing.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6Ni4zIE1vZHVsZSBhcmd1bWVudHMgYW5kIGN1c3RvbSBtb2R1bGVz">6.3 Module arguments and custom modules</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy81LjRfbW9kdWxlX2FyZ3NfY3VzdG9tX21vZHVsZS5odG1s">/workshops/5.4_module_args_custom_module.html</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6Ny4xIFBpcGVsaW5lIHZlcnNpb24gY29udHJvbCBhbmQgYWRkaXRpb25hbCByZXNvdXJjZXM=">7.1 Pipeline version control and additional resources</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLWludC1uYXZiYXI6L3dvcmtzaG9wcy81LjRfcGlwZWxpbmVfdGVzdGluZy5odG1s">/workshops/5.4_pipeline_testing.html</span></p>
</div>
<div id="quarto-meta-markdown" class="hidden">
<p><span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW1ldGF0aXRsZQ=="><strong>Nextflow Development - Samplesheet Parsing and nf-core modules</strong> – Peter Mac Nextflow Workshop</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLXR3aXR0ZXJjYXJkdGl0bGU="><strong>Nextflow Development - Samplesheet Parsing and nf-core modules</strong> – Peter Mac Nextflow Workshop</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW9nY2FyZHRpdGxl"><strong>Nextflow Development - Samplesheet Parsing and nf-core modules</strong> – Peter Mac Nextflow Workshop</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW1ldGFzaXRlbmFtZQ==">Peter Mac Nextflow Workshop</span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLXR3aXR0ZXJjYXJkZGVzYw=="></span> <span class="hidden quarto-markdown-envelope-contents" data-render-id="cXVhcnRvLW9nY2FyZGRkZXNj"></span></p>
</div>
</section>

</main> <!-- /main -->
<script id = "quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->

</body>

</html>
